packages/net/udp_socket.pony

```diff
@@ -151,16 +180,120 @@ actor UDPSocket
 
   be write(data: ByteSeq, to: NetAddress) =>
     """
-    Write a single sequence of bytes.
+    Write a single sequence of bytes. The connection will be closed if an
+    error is raised during sending.
     """
     _write(data, to)
 
   be writev(data: ByteSeqIter, to: NetAddress) =>
     """
-    Write a sequence of sequences of bytes.
+    Write a sequence of sequences of bytes as a single packet. The connection
+    will be closed if an error is raised during sending.
     """
+    if _closed then
+      return
+    end
+
+    var num_to_send: USize = 0
     for bytes in data.values() do
-      _write(bytes, to)
+      if bytes.size() == 0 then
+        // Don't send 0 byte payloads. Windows doesn't like it (and it's wasteful).
+        continue
+      end
+
+      ifdef windows then
+        _pending_sendmsg_windows.push((bytes.size(), bytes.cpointer()))
+      else
+        _pending_sendmsg_posix.push((bytes.cpointer(), bytes.size()))
+      end
+      num_to_send = num_to_send + 1
+    end
+
+    if num_to_send > 0 then
+      // Don't schedule a packet unless there's something to send.
+      _pending_packets.push((to, num_to_send))
+    end
+
+    _complete_writes()
+
+  fun ref _complete_writes() =>
+    """
+    Send pending packets. On Windows, only one packet will be sent at a time.
+    On error, dispose the connection.
+    """
+    if not _writeable then
+      return
+    end
+
+    ifdef windows then
+      if _pending_packets.size() > 0 then
```

Comment:
It seems to me, you need to loop over the _pending_packets here as done below in the non-windows case.
This might explain why there is only ever 1 packet received on the server side.

Reply:
Thanks for taking a look. My rationale for sending only a single packet at a time is that completion events might be delivered to UDPSocket out of order, so we don't know which specific entry on `_pending_packets` we're safe to free.

I missed on the docs that WSASendTo might be able to complete the operation immediately, in which case the actor will never receive a notification that it's free to continue sending. I'll try adding some logic for that, and let you know!